linkerd/app/outbound/src/opaq/concrete.rs

```diff
@@ -87,11 +94,27 @@ impl prom::encoding::EncodeLabelSet for ConcreteLabels {
     }
 }
 
-impl<T> svc::ExtractParam<balance::Metrics, Balance<T>> for BalancerMetricsParams<ConcreteLabels> {
+impl<T> svc::ExtractParam<balance::Metrics, Balance<T>> for BalancerMetricsParams<ConcreteLabels>
+where
+    T: svc::Param<Option<profiles::LogicalAddr>>,
+{
     fn extract_param(&self, bal: &Balance<T>) -> balance::Metrics {
+        // we are trying to preserve the original metrics format here.
+        // for that purpose we utilize the logical address that will be
+        // present only if we have used profiles for routing. This means:
+        //
+        // 1. If we are using policy logical and concrete will be the same
+        // 2. If we are using profiles we are doing so because the profiles
+        //    contains traffic split data. In this case the logical and
+        //    concrete are potentially different based on the backend choice.
+        let logical = match bal.parent.param() {
+            Some(profiles::LogicalAddr(addr)) => addr.to_string(),
+            None => bal.addr.to_string(),
+        };
+
         self.metrics(&ConcreteLabels {
-            logical: bal.logical.to_string().into(),
-            concrete: bal.concrete.to_string().into(),
+            logical: logical.into(),
+            concrete: bal.addr.to_string().into(),
         })
```

Comment:
I started a branch but haven't got it sensible yet, but we also need to consider the case where a balancer is provided by policy -- in that case we want to include parent and backend references (as we do in HTTP). The logical and concrete values cannot be removed due to backwards compatibility, but we should provide reference-oriented labels in the policy case to be consistent.

Reply:
@ver This makes total sense. I have put together a branch that does that. Will submit a PR agaisnt this one. So in my proposal, for the two cases we balance we get the following results. I do not see a point to special casing profiles. We can still populate all this data based on the logical addrs that we get.


#### The case when we use the **policy response** when we target a Service
```
outbound_tcp_balancer_queue_requests_total{
    parent_group="core"
    parent_kind="Service"
    parent_namespace="mixed-env"
    parent_name="legacy-app-cluster"
    parent_port="80"
    parent_section_name=""
    backend_group="core"
    backend_kind="Service"
    backend_namespace="mixed-env"
    backend_name="legacy-app-cluster"
    backend_port="80"
    backend_section_name=""
    logical="legacy-app-cluster.mixed-env.svc.cluster.local:80"
    concrete="legacy-app-cluster.mixed-env.svc.cluster.local:80"
} 1
```

#### The case when we use the **profiles** response because a `TrafficSplit` is configured
```
outbound_tcp_balancer_queue_requests_total{
    parent_group="core"
    parent_kind="Service"
    parent_namespace="trafficsplit-sample"
    parent_name="backend-svc"
    parent_port="8080"
    parent_section_name=""
    backend_group="core"
    backend_kind="Service"
    backend_namespace="trafficsplit-sample"
    backend_name="failing-svc"
    backend_port="8080"
    backend_section_name=""
    logical="backend-svc.trafficsplit-sample.svc.cluster.local:8080"
    concrete="failing-svc.trafficsplit-sample.svc.cluster.local:8080"
} 2
```